### PR TITLE
Leader sends snapshot to followers when needed

### PR DESCRIPTION
**Pull request by merlimat**
The leader will send a snapshot to followers when either: 
 * The follower is getting bootstrapped
 * The follower is tailing at an entry that was already trimmed in the leader wal

(Stacked on top of #173)

### CODE REVIEW

File: server/follower_controller.go [hunk 1]
```diff
@@ -469,15 +474,22 @@ func (fc *followerController) handleSnapshot(stream proto.OxiaLogReplication_Sen
...
+				break
+			}
 			fc.closeChannelNoMutex(err)
 			return
 		} else if snapChunk == nil {
 			break
-		} else if snapChunk.Epoch != fc.epoch {
+		} else if fc.epoch != wal.InvalidEpoch && snapChunk.Epoch != fc.epoch {
+			// The follower could be left with epoch=-1 by a previous failed
+			// attempt at sending the snapshot. It's ok to proceed int that case.
```
Comment (nahguam) [inline server/follower_controller.go]:
```suggestion
			// attempt at sending the snapshot. It's ok to proceed in that case.
```

File: server/follower_controller.go [hunk 2]
```diff
@@ -500,11 +512,31 @@ func (fc *followerController) handleSnapshot(stream proto.OxiaLogReplication_Sen
 		return
 	}
 
+	// The new epoch must be persisted, to avoid rolling it back
+	if err = newDb.UpdateEpoch(fc.epoch); err != nil {
+		fc.closeChannelNoMutex(errors.Wrap(err, "failed to update epoch in db"))
```
Comment (nahguam) [inline server/follower_controller.go]:
```suggestion
		fc.closeChannelNoMutex(errors.Wrap(err, "Failed to update epoch in db"))
```

(or the 2 instances below to lowercase)

File: server/wal/wal.go [hunk 3]
```diff
@@ -44,8 +44,13 @@ type Wal interface {
 	io.Closer
 	// Append writes an entry to the end of the log
 	Append(entry *proto.LogEntry) error
+
+	// Trim removes all the entries that are
```
Comment (nahguam) [inline server/wal/wal.go]: Looks like there should be more to the comment?

File: server/wal/wal_test.go [hunk 4]
```diff
@@ -305,3 +306,66 @@ func Clear(t *testing.T) {
 	assert.NoError(t, w.Close())
 	assert.NoError(t, f.Close())
 }
+
+func Trim(t *testing.T) {
```
Comment (nahguam) [inline server/wal/wal_test.go]:
```suggestion
func TestTrim(t *testing.T) {
```

Comment (merlimat) [reply server/wal/wal_test.go]: This is called from a main test method, instead of being a top-level test.

Comment (nahguam) [reply server/wal/wal_test.go]: ah ok

**Pull request merged**